server/src/main/java/org/apache/kafka/security/authorizer/AclEntry.java

```diff
@@ -0,0 +1,215 @@
+     * @param bytes of acls json string
+     *
+     * <p>
+        {
+            "version": 1,
+            "acls": [
+                {
+                    "host":"host1",
+                    "permissionType": "Deny",
+                    "operation": "Read",
+                    "principal": "User:alice"
+                }
+            ]
+        }
+     * </p>
+     *
+     * @return set of AclEntry objects from the JSON string
+     */
+    public static Set<AclEntry> fromBytes(byte[] bytes) throws IOException {
+        if (bytes == null || bytes.length == 0)
+            return Collections.emptySet();
+
+        Optional<JsonValue> jsonValue = Json.parseBytes(bytes);
+        if (!jsonValue.isPresent())
+            return Collections.emptySet();
+
+        JsonObject js = jsonValue.get().asJsonObject();
+
+        //the acl json version.
+        if (js.apply(VERSION_KEY).to(INT) != CURRENT_VERSION)
+            throw new IllegalArgumentException("requirement failed");
+
+        Set<AclEntry> res = new HashSet<>();
+
+        Iterator<JsonValue> aclsIter = js.apply(ACLS_KEY).asJsonArray().iterator();
+        while (aclsIter.hasNext()) {
+            JsonObject itemJs = aclsIter.next().asJsonObject();
+            KafkaPrincipal principal = SecurityUtils.parseKafkaPrincipal(itemJs.apply(PRINCIPAL_KEY).to(STRING));
+            AclPermissionType permissionType = SecurityUtils.permissionType(itemJs.apply(PERMISSION_TYPE_KEY).to(STRING));
+            String host = itemJs.apply(HOSTS_KEY).to(STRING);
+            AclOperation operation = SecurityUtils.operation(itemJs.apply(OPERATION_KEY).to(STRING));
+            res.add(AclEntry.apply(principal, permissionType, host, operation));
+        }
+
+        return res;
+    }
+
+    public static Map<String, Object> toJsonCompatibleMap(Set<AclEntry> acls) {
+        Map<String, Object> res = new HashMap<>();
+        res.put(AclEntry.VERSION_KEY, AclEntry.CURRENT_VERSION);
+        res.put(AclEntry.ACLS_KEY, acls.stream().map(AclEntry::toMap).collect(Collectors.toList()));
+        return res;
+    }
+
+    public static Set<AclOperation> supportedOperations(ResourceType resourceType) {
+        switch (resourceType) {
+            case TOPIC:
+                return new HashSet<>(Arrays.asList(READ, WRITE, CREATE, DESCRIBE, DELETE, ALTER, DESCRIBE_CONFIGS, ALTER_CONFIGS));
+            case GROUP:
+                return new HashSet<>(Arrays.asList(READ, DESCRIBE, DELETE));
+            case CLUSTER:
+                return new HashSet<>(Arrays.asList(CREATE, CLUSTER_ACTION, DESCRIBE_CONFIGS, ALTER_CONFIGS, IDEMPOTENT_WRITE, ALTER, DESCRIBE));
+            case TRANSACTIONAL_ID:
+                return new HashSet<>(Arrays.asList(DESCRIBE, WRITE));
+            case DELEGATION_TOKEN:
+                return new HashSet<>(Arrays.asList(DESCRIBE));
+            case USER:
+                return new HashSet<>(Arrays.asList(CREATE_TOKENS, DESCRIBE_TOKENS));
```

Comment:
Can't this be simple `Collections.singleton`?

Reply:
Yes. Fixed.